File: Python/find_mode_in_binary_search_tree.py
# https://leetcode.com/problems/find-mode-in-binary-search-tree/description/
from utils.examination import Examiner, Testcase
from utils.tree import TreeNode


class Solution:
    def findMode(self, root: TreeNode|None) -> list[int]:
        return []

def examine() -> None:
    Examiner(Solution(), 'findMode').run((
        Testcase((TreeNode.from_list([1,None,2,2]),), [2]),
        Testcase((TreeNode.from_list([0]),), [0]),
    ))

File: TypeScript/solutions/find-mode-in-binary-search-tree.ts
// https://leetcode.com/problems/find-mode-in-binary-search-tree/description/
import { Examiner, Testcase, TreeNode } from '../utils';


function findMode(root: TreeNode | null): number[] {
    let modes: number[] = [];
    let maxCount = 0, currentCount = 0;
    let prevVal: number | null = null;
    function inorder(node: TreeNode | null): void {
        if (node === null) return;
        inorder(node.left);
        currentCount = prevVal === node.val ? currentCount + 1 : 1;
        if (currentCount > maxCount) { maxCount = currentCount; modes = [node.val]; }
        else if (currentCount === maxCount) modes.push(node.val);
        prevVal = node.val;
        inorder(node.right);
    }
    inorder(root);
    return modes;
}

export function examine(): void {
    new Examiner(findMode).run([
        new Testcase([TreeNode.fromList([1, null, 2, 2])], [2]),
        new Testcase([TreeNode.fromList([0])], [0]),
    ]);
}
